libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/MapboxTripNotification.kt

```diff
@@ -147,31 +156,150 @@ class MapboxTripNotification(
+            currentDistanceText
+        ) { distanceFormatter, currentLegProgress, currentDistanceText ->
+            val item = currentLegProgress.currentStepProgress()?.distanceRemaining()
+            // The call below can return an empty spanable string. toString() will cause a NPE and ?. will not catch it.
+            val str = item?.let {
+                distanceFormatter.formatDistance(it)
+            }
+            if (str == null) {
+                val formattedDistance = str.toString()
+                currentDistanceText.toString() != formattedDistance
+            } else
+                false
+        } ?: false
+
+    private fun updateManeuverImage(legStep: LegStepNavigation) {
+        val maneuverImageId = getManeuverResource(legStep)
+        if (maneuverImageId != currentManeuverId) {
+            currentManeuverId = maneuverImageId
+            collapsedNotificationRemoteViews?.setImageViewResource(
+                R.id.maneuverImage,
+                maneuverImageId
+            )
+            expandedNotificationRemoteViews?.setImageViewResource(
+                R.id.maneuverImage,
+                maneuverImageId
+            )
+        }
+    }
+
+    private fun getManeuverResource(step: LegStepNavigation): Int {
+        val maneuver = step.stepManeuver()
+        val maneuverType = maneuver?.type()
+        maneuver?.let { stepManeuver ->
+            val maneuverModifier = stepManeuver.modifier()
+            if (!TextUtils.isEmpty(maneuverModifier)) {
+                val drivingSide = step.drivingSide()
+                val isLeftSideDriving = isLeftDrivingSideAndRoundaboutOrRotaryOrUturn(
+                    maneuverType,
+                    maneuverModifier,
+                    drivingSide
+                )
+                return when (isLeftSideDriving) {
+                    true -> {
+                        ManeuverResource.obtainManeuverResource(
+                            maneuverType + maneuverModifier + drivingSide
+                        )
+                    }
+                    else -> {
+                        ManeuverResource.obtainManeuverResource(
+                            maneuverType + maneuverModifier
+                        )
+                    }
+                }
+            }
+        }
+        return ManeuverResource.obtainManeuverResource(maneuverType)
+    }
+
+    private fun isLeftDrivingSideAndRoundaboutOrRotaryOrUturn(
+        maneuverType: String?,
+        maneuverModifier: String?,
+        drivingSide: String?
+    ): Boolean {
+        return STEP_MANEUVER_MODIFIER_LEFT == drivingSide &&
+            (STEP_MANEUVER_TYPE_ROUNDABOUT == maneuverType ||
+                STEP_MANEUVER_TYPE_ROTARY == maneuverType ||
+                STEP_MANEUVER_MODIFIER_UTURN == maneuverModifier
+                )
+    }
+
+    private fun onEndNavigationBtnClick() {
+        // TODO: communicate back to MapboxTripSession to stop navigation.
```

Comment:
How are you envisioning to implement the hook between `MapboxTripNotification` and `MapboxTripSession`?

Reply:
Not thought about it yet. Will work on it today